### ISSUE TITLE
Dynamic key names shall be blocked, if it would cause creation of a new key (or subkey)

### ISSUE THREAD

**Issue by test-fullautomation** [test-fullautomation/python-jsonpreprocessor#270]
Expressions on left side which contain $-operators inside double quotes shall be generally blocked.
e.g. 
`"${param}" : 1`
`"prefix_${param}_suffix" : 1`

Goal is to reduce stabilization and testing effort.
Later when all other features are stable we can create a task to provide this feature.

$-operator as part of a string as value shall be supported, if the data type is a simple data type.

**Comment by test-fullautomation** [comment]
Hi @namsonx ,
plesae consider this from now onwards for all your activities. Previous tickets may deviate or may be invalid.
Thank you,
 Thomas

**Comment by namsonx** [comment]
Hello Thomas,

I created [the commit 31b8cd50baa](https://github.com/test-fullautomation/python-jsonpreprocessor/commit/31b8cd50baaadba0f28368ba3d8444088b656ff4) to block dynamic key names, the Selftest is also updated.

Thank you,
Son

**Comment by HolQue** [comment]
Retest successful. Issue can be closed.

**Comment by HolQue** [comment]
With

`"listval" : [1,2,3],`

this works like expected:

```
"prefix_${listval}_suffix" : 1
"dictval1" : {"prefix_${listval}_suffix" : 2}
```

Result:

`Error: 'A substitution in key names is not allowed! Please update the key name "prefix_${listval}_suffix"'!`

But this does not work like expected:

```
${dictval2}['prefix_${listval}_suffix'] : 3
${dictval3.prefix_${listval}_suffix} : 4
```

Result:

`Error: 'Found expression '${listval}' with at least one parameter of composite data type ('${listval}' is of type <class 'list'>). Because of this expression is the name of a parameter, only simple data types are allowed to be substituted inside.'!
`
The last part of the error message

`"Because of this expression is the name of a parameter, only simple data types are allowed to be substituted inside."`

is invalid now. The data type is irrelevant. Because dynamic key names are principally not allowed any more.

The first error message

`Error: 'A substitution in key names is not allowed! ...`

would be the best in my opinion.

**Comment by HolQue** [comment]
Now I am not sure any more about what I wrote in my last comment.

It is possible to do this:

```
"strval"  : "A",
"dictval" : {"A_2" : 1},
${dictval}['${strval}_2'] : 2
```

Result:

`{'dictval': {'A_2': 2}, 'strval': 'A'}`

This means, it is possible to use dynamic key names. But with precondition that a parameter with resulting name does exist.

Cross check, this is not accepted:

```
"strval"   : "A",
"dictval" : {"${strval}" : 1}
```

```
"strval"   : "B",
${dictval}[${strval}] : 2
```

Result:

`Error: 'A substitution in key names is not allowed! Please update the key name "${strval}"'!`

**Conclusion:**

_To use dynamic key names is allowed, but only in case of a key with the resulting name does exist already. Therefore this mechanism can be used to overwrite the value of existing keys, but cannot be used to create new keys._

Is this correct? I need this info for the documentation.

**Comment by HolQue** [comment]
It is still possible and allowed to do this:

```
"strval"  : "A",
"dictval" : {"A_2" : 1},
${dictval}['${strval}_2'] : 2
```

Because of a key with name `"A_2"` already exists.

Result:

`{'dictval': {'A_2': 2}, 'strval': 'A'}`

But it is not allowed to do this:

```
"strval" : "B",
${dictval}['${strval}_2'] : 2
```

Because of a key with name `"B_2"` does not exist (the entire dictionary does not exist).

Result:

`Error: 'A substitution in key names is not allowed! Please update the key name "${strval}_2"'!`

_Impact: A parameter substitution inside key names is not forbidden generally. It's only forbidden in case of
a key with the resulting name does not exist already. Therefore the error message needs to be reworked._

For example:

```
Identified dynamic name of a key that does not exist. But new keys can only be created based on hard coded names.
Please remove all dollar operator expressions inside the key name "${strval}_2".
```

I will update the documentation in this way:

_To use dynamic key names is allowed, but only in case of a key with the resulting name does exist already. Therefore this mechanism can be used to overwrite the value of existing keys, but cannot be used to create new keys._

**Comment by namsonx** [comment]
Hello Holger,

Thank you for your very interesting finding! 
I think we just simply block all instances of dynamic key names, is that Okie?

Thank you, 
Son

**Comment by namsonx** [comment]
Hello Holger,

I pushed the change to stabi branch to implement the cases you mentioned in your comment.

Thank you,
Son

**Comment by HolQue** [comment]
Retest successful. Issue can be closed.

**Comment by test-fullautomation** [comment]
Released with RobotFramework AIO 0.12.0
